### PR TITLE
Make handle xml http request response protected

### PR DESCRIPTION
**Pull request by nieuwenhuisen**
## Subject

Change the visibility of the CRUDController methodes `handleXmlHttpRequestSuccessResponse` and  `handleXmlHttpRequestErrorResponse` from private to protected to allow overriding.

In my application I want to extend the Xml Http response, but there is no way to override it. The only option is to override the editAction but that's not desirable.

I am targeting this branch, because the change is not backwards compatible.

## Changelog

```markdown
### Changed
- `CRUDController::handleXmlHttpRequestSuccessResponse` method is now protected
- `CRUDController::handleXmlHttpRequestErrorResponse` method is now protected
```

### CODE REVIEW

**Comment by nieuwenhuisen** [comment]
Oke, I have changed the base branch to master

**Comment by VincentLanglet** [comment]
> These changes could not be considered BC, since the classes extending `CRUDController` could have registered the same methods with different signatures or with lower visibilities than protected.

I kinda disagree. If you consider this as a BC-break, every new function is a BC-break because we can use the same argument (if a class extend and register the function with différent signature or lower visibility...).

For instance, in https://github.com/sonata-project/SonataAdminBundle/pull/6108/files#diff-97179c4fc4120f802148ade89fc48508R342 we could have said the same.

This can be merged on 3.x IMHO.

**Comment by phansys** [comment]
>For instance, in https://github.com/sonata-project/SonataAdminBundle/pull/6108/files#diff-97179c4fc4120f802148ade89fc48508R342 we could have said the same.

I think technically both changes can lead to the same result, but the difference is here we are changing the visibility of an existing method.
IMO, there are less probabilities of reaching a collision starting from a method with a new name, than with a method with an already known name.
By instance, users generally extends the classes they need, copy and paste the methods or properties the class is requiring to work.
In this case, the most probable scenario is the users who need a custom behavior at `handleXmlHttpRequestSuccessResponse()` surely have already declared its own implementation.

**Comment by nieuwenhuisen** [comment]
I am not sure. Do I need to change anything to this PR?

**Comment by VincentLanglet** [comment]
> I am not sure. Do I need to change anything to this PR?

For me, this PR can be merge on 3.x. For @phansys, it can't be.
We need to decide with other @sonata-project/contributors

**Comment by jordisala1991** [comment]
The BC promise we are following is this one: https://symfony.com/doc/current/contributing/code/bc.html#changing-classes

Same as Symfony (which is really restrictive), and we are sometimes making exceptions (like commands)...

**About this change:** 

Private methods
Make public or protected -> Yes (It is allowed)

**About the need:**

IMO we should not open more the API. We are focusing on master on closing as much as possible.

**In conclusion:**

If I have to vote on the BC topic, it is allowed (Symfony is allowing it), but for me we shouldn't open more the api right now.

**Comment by jordisala1991** [comment]
Maybe it is not super nice, but in this case, can't you override the renderJson method? or it is private too?

**Comment by VincentLanglet** [comment]
> IMO we should not open more the API. We are focusing on master on closing as much as possible.

I would say that we're trying to close if possible.
But in multiple case, it make sens to me that method are protected instead of private.
For instance, we recently merged this PR: https://github.com/sonata-project/SonataAdminBundle/pull/5965

If someone wants to use our code, with good reason, I think she should always provide an easy way to do it.

In which way you want to override `handleXmlHttpRequestSuccessResponse` and `handleXmlHttpRequestErrorResponse` @nieuwenhuisen ?

**Comment by nieuwenhuisen** [comment]
I have changed the branch to 3.x

> In which way you want to override handleXmlHttpRequestSuccessResponse and handleXmlHttpRequestErrorResponse @nieuwenhuisen ?

I want to add an extra field with the full html render of the form. This makes it easy to fully replace the form with the correct error's visible. We need AJAX for this because we want to update a couple form at the same time.

**Comment by core23** [comment]
Can you please rebase, otherwise the CI system won't run and we can't merge your changes @nieuwenhuisen

**Comment by phansys** [comment]
Forcing the pipeline tasks to be executed again.

File: src/Controller/CRUDController.php [hunk 1]
```diff
@@ -1520,6 +1520,41 @@ protected function preList(Request $request)
         return $this->get('translator')->trans($id, $parameters, $domain, $locale);
     }
 
+    protected function handleXmlHttpRequestErrorResponse(Request $request, FormInterface $form): ?JsonResponse
+    {
+        if (empty(array_intersect(['application/json', '*/*'], $request->getAcceptableContentTypes()))) {
+            @trigger_error('In next major version response will return 406 NOT ACCEPTABLE without `Accept: application/json` or `Accept: */*`', E_USER_DEPRECATED);
```
Comment (phansys) [inline src/Controller/CRUDController.php:1526]: This message should be explicit about the version where the absence of the header was deprecated.

Comment (VincentLanglet) [reply src/Controller/CRUDController.php:1526]:
```
Foo is not supported since sonata-projet/admin-bundle 3.x and will return 406 in 4.0
```

**Comment by VincentLanglet** [comment]
Are you still interested in this PR @nieuwenhuisen ?

File: src/Controller/CRUDController.php (hunk 1 above)
Comment (nieuwenhuisen) [reply src/Controller/CRUDController.php:1538]: The PR is about making the methodes protected, this line is not changed. Is it not better to do this change in a separate PR to keep the git commit etc relevant?

**Comment by nieuwenhuisen** [comment]
Yes, I will do a rebase this week

File: src/Controller/CRUDController.php (hunk 1 above)
Comment (VincentLanglet) [reply src/Controller/CRUDController.php:1538]: Oh, it's not a new deprecation indeed.
We should find since when it's deprecated and modify the message, and add a NEXT_MAJOR comment.
It can be done in another PR or in another commit.

Comment (nieuwenhuisen) [reply src/Controller/CRUDController.php]: I have change the deprecation message in a new commit.

File: src/Controller/CRUDController.php [hunk 2]
```diff
@@ -1558,7 +1558,7 @@ protected function preList(Request $request)
     protected function handleXmlHttpRequestErrorResponse(Request $request, FormInterface $form): ?JsonResponse
     {
         if (empty(array_intersect(['application/json', '*/*'], $request->getAcceptableContentTypes()))) {
-            @trigger_error('In next major version response will return 406 NOT ACCEPTABLE without `Accept: application/json` or `Accept: */*`', E_USER_DEPRECATED);
+            @trigger_error(sprintf('%s is not supported since sonata-project/admin-bundle 3.x and will return 406 in 4.0. Use `application/json`.', implode(',', $request->getAcceptableContentTypes())));
```
Comment (phansys) [inline src/Controller/CRUDController.php:1561]:
```suggestion
            @trigger_error(sprintf('%s is not supported since sonata-project/admin-bundle 3.x and will return 406 in 4.0. Use `application/json`.', implode(',', $request->getAcceptableContentTypes())));
```
```suggestion
            @trigger_error(sprintf('"%s" is not supported since sonata-project/admin-bundle 3.x and will return 406 in 4.0. Use `application/json`.', implode('", "', $request->getAcceptableContentTypes())));
```
IMO, this message lacks the context (request, XHR, etc) . If this is shown in the deprecation logs, I think it will be difficult to understand where it was triggered and what should be done to fix the problem.

Comment (phansys) [reply src/Controller/CRUDController.php:1561]: BTW, we should also be more explicit about this message:
>will return 406

 I propose something like:
>the status code 406 will be set in the returned response

File: src/Controller/CRUDController.php [hunk 3]
```diff
@@ -1555,6 +1555,41 @@ protected function preList(Request $request)
...
+        return $this->renderJson([
+            'result' => 'error',
+            'errors' => $errors,
+        ], Response::HTTP_BAD_REQUEST);
+    }
+
+    /**
+     * @param object $object
+     */
+    protected function handleXmlHttpRequestSuccessResponse(Request $request, $object): JsonResponse
```
Comment (core23) [inline src/Controller/CRUDController.php:1580]:
```suggestion
    protected function handleXmlHttpRequestSuccessResponse(Request $request, object $object): JsonResponse
```

**Review by core23 — CHANGES_REQUESTED** [review]

File: src/Controller/CRUDController.php (hunk 3 above)
Comment (nieuwenhuisen) [reply src/Controller/CRUDController.php]: Aangepast

**Review by core23 — APPROVED** [review]

File: src/Controller/CRUDController.php [hunk 4]
```diff
@@ -1555,6 +1555,38 @@ protected function preList(Request $request)
         return $this->get('translator')->trans($id, $parameters, $domain, $locale);
     }
 
+    protected function handleXmlHttpRequestErrorResponse(Request $request, FormInterface $form): ?JsonResponse
+    {
+        if (empty(array_intersect(['application/json', '*/*'], $request->getAcceptableContentTypes()))) {
+            @trigger_error(sprintf('"%s" is not supported since sonata-project/admin-bundle 3.x and the status code 406 will be set in the returned response in 4.0. Use `application/json`.', implode('", "', $request->getAcceptableContentTypes())), E_USER_DEPRECATED);
```
Comment (phansys) [inline src/Controller/CRUDController.php:1561]: This is just a suggestion. I also think that this message could be improved even more in cases where the "Accept" header is omitted, in order to avoid showing `""` as the passed value.

```suggestion
            @trigger_error(sprintf(
                'None of the passed values ("%s") in the "Accept" header when requesting %s %s is supported since sonata-project/admin-bundle 3.x.'
                .' It will result in a response with the status code 406 (Not Acceptable) in 4.0. You must add `application/json`.',
                implode('", "', $request->getAcceptableContentTypes()),
                $request->getMethod(),
                $request->getUri()
            ), E_USER_DEPRECATED);
```

**Review by phansys — APPROVED** [review]

**Review by phansys — APPROVED** [review]

**Review by core23 — APPROVED** [review]

**Comment by VincentLanglet** [comment]
Thanks @nieuwenhuisen

**Comment by VincentLanglet** [comment]
Hi @nieuwenhuisen, can you give an example about how you override the method.
We're trying to close the API for Sonata 4 and I'd like to see if there is another solution for your use case.
